### ISSUE TITLE
CUE_EXPERIMENT=toposort Feedback

### ISSUE THREAD

**Issue by cuematthew** [cue-lang/cue#3558]
By running `cue eval` and `cue export` with `CUE_EXPERIMENT=toposort`, you enable the new "topological sorting" algorithm. This can be combined with other experimental flags, e.g. `CUE_EXPERIMENT=toposort,evalv3`.

EDIT: Since v0.12.0-alpha.2, "toposort" is enabled by default. If you need to turn it off, set `CUE_EXPERIMENT=toposort=0` in your environment.

The intention is that "topological sorting" should produce the same output with both versions of the evaluator, though we are aware of some minor differences at the time of writing.

It is called "topological sorting" because it creates a graph of field names and attempts to find a topological walk through that graph. For example:

```
x: {g: _, b: _, c: _}
y: {c: _, a: _}
z: x & y
```

For `z`, we establish a graph with the following edges:
  - `g` before `b`
  - `b` before `c`
  - `c` before `a`
 
 From these edges, there's a simple topological walk, which guarantees the output of `z` will be:

```
{g: _, b: _, c: _, a: _}
```

However, cycles can exist in this graph. Consider:

```
x: {g: _, b: _, c: _}
y: {c: _, a: _, b: _}
z: x & y
```

This minor change to the original example adds in the edge `a` before `b`, thus creating a cycle in `z`. Our algorithm detects these cycles and attempts to find a sensible place to break such cycles, in a deterministic fashion. Currently, in this particular scenario, after reaching `g`, it then finds the cycle (`a -> b -> c -> a`) and chooses `b` as the entry to the cycle (which is justified because of the edge `g -> b`, and the fact there's no edge from `g` (or any predecessor of `g`) to any other node within the cycle), so the output of `z` will be `{g: _, b: _, c: _, a: _}`.

Please use this issue to record further feedback; surprises, undesirable behaviour etc.

**Comment by mxey** [comment]
When calling json.Marshal on the same struct twice during a CUE run, should the output be identical? https://github.com/cue-lang/cue/issues/3969

**Comment by mvdan** [comment]
@mxey it depends on what you mean by "same struct"; please provide a small example here because the examples at #3969 are quite large.

If you mean "same struct value", the answer is not necessarily: https://cuelang.org/play/?id=t9jKwaplo0E#w=function&i=cue&f=export&o=cue

If you mean "unify with the same struct reference", again the answer is not necessarily: https://cuelang.org/play/?id=Grjh_YxJfnN#w=function&i=cue&f=export&o=cue

**Comment by mxey** [comment]
@mvdan yeah I‘ll have to create a simpler example on Monday.  

What I mean is not the same fields and values, but the very same struct, like

```
s: {a: 1, b: 2}
a: json.Marshal(s)
b: json.Marshal(s)
````

I would think a and b should always be identical then.

**Comment by cuematthew** [comment]
(I've not caught up with @mvdan yet this morning, so I might well be missing key context here)

> I would think a and b should always be identical then.

I agree with you. From this example though, I can't provoke it behaving differently.
The ordering of fields within `s` should be deterministic. If it's not, that's 100% a bug, and a bad one too.

**Comment by mxey** [comment]
@cuematthew yeah, the simple example does not exhibit the wrong behaviour, I only used it to make sure we agreed on the invariant. There is a full write-up of my situation at https://github.com/cue-lang/cue/issues/3969#issuecomment-2975482264

**Comment by cuematthew** [comment]
Thanks @mxey - I'm (slowly) getting up to speed on this now.

**Comment by mvdan** [comment]
As of the release of v0.14.0, `CUE_EXPERIMENT=toposort` is always on and can no longer be disabled. Hence, given that the experiment is concluded, I think we can close this issue now, as its purpose was for users of the experiment to share early feedback.

If you run into any further issues, please file a new issue or message on Slack or Discord. Thanks!
